# Data Analysis
# Practical 7

# Q1 Libraries

In [17]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Q2 Dataset

In [18]:
penguins = pd.read_csv('penguins_size.csv')
penguins.head(10)

species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen              39.1             18.7              181.0   
1  Adelie  Torgersen              39.5             17.4              186.0   
2  Adelie  Torgersen              40.3             18.0              195.0   
3  Adelie  Torgersen               NaN              NaN                NaN   
4  Adelie  Torgersen              36.7             19.3              193.0   
5  Adelie  Torgersen              39.3             20.6              190.0   
6  Adelie  Torgersen              38.9             17.8              181.0   
7  Adelie  Torgersen              39.2             19.6              195.0   
8  Adelie  Torgersen              34.1             18.1              193.0   
9  Adelie  Torgersen              42.0             20.2              190.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  
5       3650.0    MALE  
6       3625.0  FEMALE  
7       4675.0    MALE  
8       3475.0     NaN  
9       4250.0     NaN

In [19]:
penguins.shape

(344, 7)

# Q3 Sum of null/missing values in the dataset for each column at once

In [20]:
peng_nans = penguins.isna()
peng_nans.sum()

species               0
island                0
culmen_length_mm      2
culmen_depth_mm       2
flipper_length_mm     2
body_mass_g           2
sex                  10
dtype: int64

In [21]:
peng_nans

species  island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0      False   False             False            False              False   
1      False   False             False            False              False   
2      False   False             False            False              False   
3      False   False              True             True               True   
4      False   False             False            False              False   
..       ...     ...               ...              ...                ...   
339    False   False              True             True               True   
340    False   False             False            False              False   
341    False   False             False            False              False   
342    False   False             False            False              False   
343    False   False             False            False              False   

     body_mass_g    sex  
0          False  False  
1          False  False  
2          False  False  
3           True   True  
4          False  False  
..           ...    ...  
339         True   True  
340        False  False  
341        False  False  
342        False  False  
343        False  False  

[344 rows x 7 columns]

# Q4 Total number of missing values

In [22]:
print(peng_nans.sum().sum())

18


# Q5 Fill the missing values

In [23]:
peng_filled = penguins.bfill(axis=1)
# df = df.apply(lambda row: row.fillna(method='bfill'), axis=1)
peng_filled

species     island culmen_length_mm culmen_depth_mm flipper_length_mm  \
0    Adelie  Torgersen             39.1            18.7             181.0   
1    Adelie  Torgersen             39.5            17.4             186.0   
2    Adelie  Torgersen             40.3            18.0             195.0   
3    Adelie  Torgersen              NaN             NaN               NaN   
4    Adelie  Torgersen             36.7            19.3             193.0   
..      ...        ...              ...             ...               ...   
339  Gentoo     Biscoe              NaN             NaN               NaN   
340  Gentoo     Biscoe             46.8            14.3             215.0   
341  Gentoo     Biscoe             50.4            15.7             222.0   
342  Gentoo     Biscoe             45.2            14.8             212.0   
343  Gentoo     Biscoe             49.9            16.1             213.0   

    body_mass_g     sex  
0        3750.0    MALE  
1        3800.0  FEMALE  
2        3250.0  FEMALE  
3           NaN     NaN  
4        3450.0  FEMALE  
..          ...     ...  
339         NaN     NaN  
340      4850.0  FEMALE  
341      5750.0    MALE  
342      5200.0  FEMALE  
343      5400.0    MALE  

[344 rows x 7 columns]

In [24]:
print(peng_filled.isnull().sum().sum())

18


### I am not sure what went wrong, but I still see missing values

# Q6 Fill the miss. values with values in left column

In [25]:
peng_left_filled = penguins.apply(lambda row: row.ffill(), axis=1)
peng_left_filled

species     island culmen_length_mm culmen_depth_mm flipper_length_mm  \
0    Adelie  Torgersen             39.1            18.7             181.0   
1    Adelie  Torgersen             39.5            17.4             186.0   
2    Adelie  Torgersen             40.3            18.0             195.0   
3    Adelie  Torgersen        Torgersen       Torgersen         Torgersen   
4    Adelie  Torgersen             36.7            19.3             193.0   
..      ...        ...              ...             ...               ...   
339  Gentoo     Biscoe           Biscoe          Biscoe            Biscoe   
340  Gentoo     Biscoe             46.8            14.3             215.0   
341  Gentoo     Biscoe             50.4            15.7             222.0   
342  Gentoo     Biscoe             45.2            14.8             212.0   
343  Gentoo     Biscoe             49.9            16.1             213.0   

    body_mass_g        sex  
0        3750.0       MALE  
1        3800.0     FEMALE  
2        3250.0     FEMALE  
3     Torgersen  Torgersen  
4        3450.0     FEMALE  
..          ...        ...  
339      Biscoe     Biscoe  
340      4850.0     FEMALE  
341      5750.0       MALE  
342      5200.0     FEMALE  
343      5400.0       MALE  

[344 rows x 7 columns]

In [26]:
print(peng_left_filled.isna().sum())

species              0
island               0
culmen_length_mm     0
culmen_depth_mm      0
flipper_length_mm    0
body_mass_g          0
sex                  0
dtype: int64


### Ok, what even is the idea behind this thing? What's the point of filling like this? This makes no sense

# Q7 Sex column

In [27]:
peng_sex_filled = penguins['sex'].fillna('FEMALE')

In [52]:
peng_sex_filled.isna().sum()

np.int64(0)

In [ ]:
peng_sex_filled['sex'].head(10)

0      MALE
1    FEMALE
2    FEMALE
3    FEMALE
4    FEMALE
5      MALE
6    FEMALE
7      MALE
8    FEMALE
9    FEMALE
Name: sex, dtype: object

# Q8 Filling bill_length_mm with max(culmen_length_mm in my case)

### I think I chose the wrong dataset, but I could still use culmen_length_mm column

In [ ]:
peng_culmen_length_filled = penguins['culmen_length_mm'].fillna(penguins['culmen_length_mm'].max())

In [ ]:
peng_culmen_length_filled['culmen_length_mm'].isna().sum()

np.int64(0)

In [ ]:
peng_culmen_length_filled['culmen_length_mm'].head(10)

0    39.1
1    39.5
2    40.3
3    59.6
4    36.7
5    39.3
6    38.9
7    39.2
8    34.1
9    42.0
Name: culmen_length_mm, dtype: float64

# Q9 Filling the same column with mean values

In [30]:
peng_cl_mean_filled = penguins['culmen_length_mm'].fillna(penguins['culmen_length_mm'].mean())

In [31]:
peng_cl_mean_filled.isna().sum()

np.int64(0)

In [32]:
peng_cl_mean_filled.head(10)

0    39.10000
1    39.50000
2    40.30000
3    43.92193
4    36.70000
5    39.30000
6    38.90000
7    39.20000
8    34.10000
9    42.00000
Name: culmen_length_mm, dtype: float64

### As you can see, the element at 3rd index was NaN initially, when we used max value, it took the value fof 59.6, now it takes the value of 43.921. I think this value is better. Because it quite 'fits' the dataset

# Q10 Interpolation

In [34]:
penguins_interpolated = penguins.interpolate(method='linear')

C:\Users\n.sharmetov\AppData\Local\Temp\ipykernel_26908\65074680.py:1: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  penguins_interpolated = penguins.interpolate(method='linear')


In [35]:
penguins_interpolated.isna().sum()

species               0
island                0
culmen_length_mm      0
culmen_depth_mm       0
flipper_length_mm     0
body_mass_g           0
sex                  10
dtype: int64

### Sex remains unchanged, becaus it is qualitative

# Q11 Interpolation to 'flipper_length_mm' column

In [36]:
peng_fl_mm_interpol = penguins['flipper_length_mm'].interpolate(method='linear')

In [39]:
peng_fl_mm_interpol.isna().sum()

np.int64(0)

In [41]:
peng_fl_mm_interpol.head(10)

0    181.0
1    186.0
2    195.0
3    194.0
4    193.0
5    190.0
6    181.0
7    195.0
8    193.0
9    190.0
Name: flipper_length_mm, dtype: float64

# Q12 Drop all null values

In [42]:
peng_no_nans = penguins.dropna()

In [43]:
peng_no_nans.isna().sum()

species              0
island               0
culmen_length_mm     0
culmen_depth_mm      0
flipper_length_mm    0
body_mass_g          0
sex                  0
dtype: int64

In [44]:
peng_no_nans.head(10)

species     island  culmen_length_mm  culmen_depth_mm  flipper_length_mm  \
0   Adelie  Torgersen              39.1             18.7              181.0   
1   Adelie  Torgersen              39.5             17.4              186.0   
2   Adelie  Torgersen              40.3             18.0              195.0   
4   Adelie  Torgersen              36.7             19.3              193.0   
5   Adelie  Torgersen              39.3             20.6              190.0   
6   Adelie  Torgersen              38.9             17.8              181.0   
7   Adelie  Torgersen              39.2             19.6              195.0   
12  Adelie  Torgersen              41.1             17.6              182.0   
13  Adelie  Torgersen              38.6             21.2              191.0   
14  Adelie  Torgersen              34.6             21.1              198.0   

    body_mass_g     sex  
0        3750.0    MALE  
1        3800.0  FEMALE  
2        3250.0  FEMALE  
4        3450.0  FEMALE  
5        3650.0    MALE  
6        3625.0  FEMALE  
7        4675.0    MALE  
12       3200.0  FEMALE  
13       3800.0    MALE  
14       4400.0    MALE

In [45]:
peng_no_nans.shape

(334, 7)

In [46]:
penguins.shape

(344, 7)

In [47]:
10/344

0.029069767441860465

### OK, 10 out of 344 (2%) values were dropped

# Q13 Group species and count values

In [49]:
grouping = peng_no_nans['species'].groupby(peng_no_nans['island']).value_counts()

In [50]:
grouping

island     species  
Biscoe     Gentoo       120
           Adelie        44
Dream      Chinstrap     68
           Adelie        55
Torgersen  Adelie        47
Name: count, dtype: int64

# Q14 Continue Q13 and build cross tabulation

In [51]:
pd.crosstab(peng_no_nans['species'], peng_no_nans['island'], normalize='all')

island       Biscoe     Dream  Torgersen
species                                 
Adelie     0.131737  0.164671   0.140719
Chinstrap  0.000000  0.203593   0.000000
Gentoo     0.359281  0.000000   0.000000

### This clear structure makes it visible that Biscoe species from Gentoo island is most dominant in this dataset taking 35% of it. I bellive, even if we did not remove the NaN values, the result will be extremely similar, because we only removed 2 % of the dataset